In [19]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error
import pandas as pd
df = pd.read_csv("IS310 DATA - Sheet1.csv")
df

game_name   YOR operability status  \
0                                    Bring It On  2005           operable   
1                            Alice in Wonderland  2007           operable   
2                        Anime Character Maker 2  2009           operable   
3                                Beautiful Alice  2008           operable   
4                            Beach Doll Dress Up  2006       non-operable   
5                             DKNY Spring 2008 2  2008           operable   
6   [S] June: Star in a Dress-Up Game Circa 2009  2009           operable   
7              50's Fashion Dress Up Game (2010)  2010           operable   
8                    Cover Model Dress Up: April  2010       non-operable   
9                        2010 Teens Fashion Show  2010           operable   
10                    Anime Bikini Dress Up Game  2012           operable   
11             284 Modern Princess Dress Up Game  2013       non-operable   
12                      Summer BFF Dress Up Game  2014       non-operable   
13                        Amazon Warrior Creator  2018           operable   
14                    Cinderella Flies to Mexico  2015       non-operable   
15                               Color Me Lolita  2015       non-operable   
16              Anime Summer Girls Dress Up Game  2016           operable   
17                     Boho Chic Spring Shopping  2017       non-operable   
18                         Swimming Pool Delight  2018           operable   
19                          Geisha Dress-Up Game  2018       non-operable   

                  DEVELOPER                  PUBLISHER GENDER  \
0                    G-Game             \tYahoo! Korea      F   
1                DressUpWho                 DressUpWho      F   
2                   Xdanond                 Kongregate    M/F   
3               AllForGirls                 Kongregate      F   
4          www.roiworld.com           www.roiworld.com      F   
5   Glorious Games Group AB                        NaN      F   
6             lavendersiren                 \tStardoll      F   
7                Pichichama  DeviantArt; Rinmaru Games      F   
8                Spil Games           girlsgogames.com      F   
9               iGirlsGames                iGirlsGames      F   
10               Pichichama  DeviantArt; Rinmaru Games      F   
11               Pichichama  DeviantArt; Rinmaru Games      F   
12               Pichichama              Rinmaru Games      F   
13               Pichichama           rinmarugames.com      F   
14               DressUpWho                 DressUpWho      F   
15                   Lina01                     Lina01      F   
16         PrinceofRedRoses           PrinceofRedRoses      F   
17          Games for Girlz            Games for Girlz      F   
18                GirlGames                  GirlGames      F   
19             AzaleasDolls                 DeviantArt      F   

    NO. OF SKINTONES                                          GAME LINK  
0                  1  https://ooooooooo.ooo/static/?e0f637b9-e5bd-41...  
1                  1  https://ooooooooo.ooo/static/?2bb5ed50-cf95-40...  
2                 11  https://ooooooooo.ooo/static/?dba0d4ce-0747-85...  
3                  1  https://ooooooooo.ooo/static/?c65c45ab-7375-42...  
4                  1  https://ooooooooo.ooo/static/?92db83d9-1dce-49...  
5                  1  http://cdn.stardoll.com/flash/loader.swf?real_...  
6                  1  http://pipe.miroware.io/609e0655a333fb12a3acf5...  
7                  4  https://www.deviantart.com/pichichama/art/50-s...  
8                  2  https://flashpointproject.github.io/flashpoint...  
9                  1  https://flashpointproject.github.io/flashpoint...  
10                 6  https://flashpointproject.github.io/flashpoint...  
11                11  https://flashpointproject.github.io/flashpoint...  
12                 9  https://flashpointproject.github.io/flashpoint...  
13                1

In [20]:
df_pre_2011 = df[df['YOR'] < 2011]
df_post_2011 = df[df['YOR'] >= 2011]

X_pre_2011 = df_pre_2011[['YOR']]
y_pre_2011 = df_pre_2011['NO. OF SKINTONES']

model_pre_2011 = LinearRegression()
model_pre_2011.fit(X_pre_2011, y_pre_2011)

df_pre_2011['predicted_skin_tones'] = model_pre_2011.predict(X_pre_2011)

X_post_2011 = df_post_2011[['YOR']]
y_post_2011 = df_post_2011['NO. OF SKINTONES']

model_post_2011 = LinearRegression()
model_post_2011.fit(X_post_2011, y_post_2011)

df_post_2011['predicted_skin_tones'] = model_post_2011.predict(X_post_2011)
df_combined = pd.concat([df_pre_2011, df_post_2011])
df_combined[['game_name', 'YOR', 'NO. OF SKINTONES', 'predicted_skin_tones']]

/var/folders/s5/wltcb2kn14q_f2mvntm4xf500000gn/T/ipykernel_20438/2801561373.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_pre_2011['predicted_skin_tones'] = model_pre_2011.predict(X_pre_2011)
/var/folders/s5/wltcb2kn14q_f2mvntm4xf500000gn/T/ipykernel_20438/2801561373.py:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_post_2011['predicted_skin_tones'] = model_post_2011.predict(X_post_2011)


game_name   YOR  NO. OF SKINTONES  \
0                                    Bring It On  2005                 1   
1                            Alice in Wonderland  2007                 1   
2                        Anime Character Maker 2  2009                11   
3                                Beautiful Alice  2008                 1   
4                            Beach Doll Dress Up  2006                 1   
5                             DKNY Spring 2008 2  2008                 1   
6   [S] June: Star in a Dress-Up Game Circa 2009  2009                 1   
7              50's Fashion Dress Up Game (2010)  2010                 4   
8                    Cover Model Dress Up: April  2010                 2   
9                        2010 Teens Fashion Show  2010                 1   
10                    Anime Bikini Dress Up Game  2012                 6   
11             284 Modern Princess Dress Up Game  2013                11   
12                      Summer BFF Dress Up Game  2014                 9   
13                        Amazon Warrior Creator  2018                16   
14                    Cinderella Flies to Mexico  2015                 1   
15                               Color Me Lolita  2015                 1   
16              Anime Summer Girls Dress Up Game  2016                14   
17                     Boho Chic Spring Shopping  2017                 6   
18                         Swimming Pool Delight  2018                 5   
19                          Geisha Dress-Up Game  2018                 5   

    predicted_skin_tones  
0               0.637681  
1               1.739130  
2               2.840580  
3               2.289855  
4               1.188406  
5               2.289855  
6               2.840580  
7               3.391304  
8               3.391304  
9               3.391304  
10              6.500000  
11              6.750000  
12              7.000000  
13              8.000000  
14              7.250000  
15              7.250000  
16              7.500000  
17              7.750000  
18              8.000000  
19              8.000000

In [21]:
df_pre_2011

game_name   YOR operability status  \
0                                   Bring It On  2005           operable   
1                           Alice in Wonderland  2007           operable   
2                       Anime Character Maker 2  2009           operable   
3                               Beautiful Alice  2008           operable   
4                           Beach Doll Dress Up  2006       non-operable   
5                            DKNY Spring 2008 2  2008           operable   
6  [S] June: Star in a Dress-Up Game Circa 2009  2009           operable   
7             50's Fashion Dress Up Game (2010)  2010           operable   
8                   Cover Model Dress Up: April  2010       non-operable   
9                       2010 Teens Fashion Show  2010           operable   

                 DEVELOPER                  PUBLISHER GENDER  \
0                   G-Game             \tYahoo! Korea      F   
1               DressUpWho                 DressUpWho      F   
2                  Xdanond                 Kongregate    M/F   
3              AllForGirls                 Kongregate      F   
4         www.roiworld.com           www.roiworld.com      F   
5  Glorious Games Group AB                        NaN      F   
6            lavendersiren                 \tStardoll      F   
7               Pichichama  DeviantArt; Rinmaru Games      F   
8               Spil Games           girlsgogames.com      F   
9              iGirlsGames                iGirlsGames      F   

   NO. OF SKINTONES                                          GAME LINK  \
0                 1  https://ooooooooo.ooo/static/?e0f637b9-e5bd-41...   
1                 1  https://ooooooooo.ooo/static/?2bb5ed50-cf95-40...   
2                11  https://ooooooooo.ooo/static/?dba0d4ce-0747-85...   
3                 1  https://ooooooooo.ooo/static/?c65c45ab-7375-42...   
4                 1  https://ooooooooo.ooo/static/?92db83d9-1dce-49...   
5                 1  http://cdn.stardoll.com/flash/loader.swf?real_...   
6                 1  http://pipe.miroware.io/609e0655a333fb12a3acf5...   
7                 4  https://www.deviantart.com/pichichama/art/50-s...   
8                 2  https://flashpointproject.github.io/flashpoint...   
9                 1  https://flashpointproject.github.io/flashpoint...   

   predicted_skin_tones  
0              0.637681  
1              1.739130  
2              2.840580  
3              2.289855  
4              1.188406  
5              2.289855  
6              2.840580  
7              3.391304  
8              3.391304  
9              3.391304

In [22]:
df_post_2011

game_name   YOR operability status  \
10         Anime Bikini Dress Up Game  2012           operable   
11  284 Modern Princess Dress Up Game  2013       non-operable   
12           Summer BFF Dress Up Game  2014       non-operable   
13             Amazon Warrior Creator  2018           operable   
14         Cinderella Flies to Mexico  2015       non-operable   
15                    Color Me Lolita  2015       non-operable   
16   Anime Summer Girls Dress Up Game  2016           operable   
17          Boho Chic Spring Shopping  2017       non-operable   
18              Swimming Pool Delight  2018           operable   
19               Geisha Dress-Up Game  2018       non-operable   

           DEVELOPER                  PUBLISHER GENDER  NO. OF SKINTONES  \
10        Pichichama  DeviantArt; Rinmaru Games      F                 6   
11        Pichichama  DeviantArt; Rinmaru Games      F                11   
12        Pichichama              Rinmaru Games      F                 9   
13        Pichichama           rinmarugames.com      F                16   
14        DressUpWho                 DressUpWho      F                 1   
15            Lina01                     Lina01      F                 1   
16  PrinceofRedRoses           PrinceofRedRoses      F                14   
17   Games for Girlz            Games for Girlz      F                 6   
18         GirlGames                  GirlGames      F                 5   
19      AzaleasDolls                 DeviantArt      F                 5   

                                            GAME LINK  predicted_skin_tones  
10  https://flashpointproject.github.io/flashpoint...                  6.50  
11  https://flashpointproject.github.io/flashpoint...                  6.75  
12  https://flashpointproject.github.io/flashpoint...                  7.00  
13  https://flashpointproject.github.io/flashpoint...                  8.00  
14  https://flashpointproject.github.io/flashpoint...                  7.25  
15  https://flashpointproject.github.io/flashpoint...                  7.25  
16  https://flashpointproject.github.io/flashpoint...                  7.50  
17  https://flashpointproject.github.io/flashpoint...                  7.75  
18  https://flashpointproject.github.io/flashpoint...                  8.00  
19  https://flashpointproject.github.io/flashpoint...                  8.00

In [23]:
mae_pre_2011 = mean_absolute_error(df_pre_2011['NO. OF SKINTONES'], df_pre_2011['predicted_skin_tones'])
mae_post_2011 = mean_absolute_error(df_post_2011['NO. OF SKINTONES'], df_post_2011['predicted_skin_tones'])
metrics = {
    "Pre-2011 Model": {
        "MAE": mae_pre_2011,
    },
    "Post-2011 Model": {
        "MAE": mae_post_2011,
    }
}

metrics_df = pd.DataFrame(metrics).T
metrics_df

MAE
Pre-2011 Model   1.826087
Post-2011 Model  4.150000

Measuring Accuracy:

This is the average of the absolute differences between the predicted values and the actual values. It measures the magnitude of errors in a set of predictions, without considering their direction (positive or negative).

Interpretation:

Pre-2011 Model: The MAE is 1.83, meaning that, on average, the predicted number of skin tones is off by about 1.83 skin tones from the actual values.
Post-2011 Model: The MAE is 4.15, which means the predictions are, on average, off by about 4.15 skin tones.

1. Understanding Trends Over Time:

Pre-2011 vs. Post-2011 Trends: By analyzing the regression models for pre-2011 and post-2011 games, we can see how skin tone diversity has shifted over the years. The analysis shows that games from the pre-2011 era generally had fewer skin tones, with predictions for the post-2011 period showing an increase in diversity. This supports the idea that societal shifts toward inclusivity and diversity likely influenced game developers to offer more skin tones in later years.

2. Eliminating Manual Data Entry:

The prediction model allows us to instantly predict the number of skin tones for any game, based on its year of release. This speeds up the entire data collection process, especially when dealing with large datasets or numerous games. For instance, if we were previously analyzing hundreds of games, manually counting skin tones for each would take considerable time. The model allows for quicker analysis and the ability to process large volumes of data in much less time.